File: summary/c++/17_shakutori.cpp
#include <iostream>
#include <vector>

int shakutori(const std::vector<int>& arr, int k) {
    int n = arr.size();
    int res = 0;
    int right = 0;
    int sum = 0;

    for (int left = 0; left < n; ++left) {
        while (right < n && sum + arr[right] <= k) {
            sum += arr[right];
            ++right;
        }

        res += right - left;

        if (right == left) {
            ++right;
        } else {
            sum -= arr[left];
        }
    }

    return res;
}

int main() {
    int n, k;
    std::cin >> n >> k;
    std::vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    int result = shakutori(arr, k);
    std::cout << "Number of subarrays with sum <= " << k << ": " << result << std::endl;

    return 0;
}

File: summary/dart/17_shakutori.dart
void main() {
  List<int> nums = [1, 2, 3, 4, 2, 5, 1, 2, 3];
  int target = 7;
  List<int> result = longestSubarrayWithSum(nums, target);
  print(result); // 出力例: [3, 4]
}

List<int> longestSubarrayWithSum(List<int> nums, int target) {
  int left = 0;
  int right = 0;
  int currentSum = 0;
  int maxLength = 0;
  int startIndex = 0;

  while (right < nums.length) {
    currentSum += nums[right];
    while (currentSum > target && left <= right) {
      currentSum -= nums[left];
      left++;
    }

    if (currentSum == target && (right - left + 1) > maxLength) {
      maxLength = right - left + 1;
      startIndex = left;
    }

    right++;
  }

  return nums.sublist(startIndex, startIndex + maxLength);
}
